File: 1160-letter-tile-possibilities/letter-tile-possibilities.py
class Solution:
    def numTilePossibilities(self, tiles: str) -> int:
        from itertools import permutations
        
        # Use a set to store unique sequences
        unique_sequences = set()
        
        # Iterate over all possible lengths of sequences
        for length in range(1, len(tiles) + 1):
            # Generate all permutations of the current length
            for perm in permutations(tiles, length):
                # Add the permutation to the set
                unique_sequences.add(perm)
        
        # The number of unique sequences is the size of the set
        return len(unique_sequences)

File: 1160-letter-tile-possibilities/letter-tile-possibilities.java
class Solution {
    public int numTilePossibilities(String tiles) {
        // Use a Set to store unique sequences
        Set<String> uniqueSequences = new HashSet<>();
        
        // Generate all possible permutations for all lengths
        generatePermutations(tiles.toCharArray(), 0, uniqueSequences);
        
        // The number of unique sequences is the size of the set
        return uniqueSequences.size();
    }
    private void generatePermutations(char[] tiles, int index, Set<String> uniqueSequences) {
        if (index >= tiles.length) {
            return;
        }

        for (int i = index; i < tiles.length; i++) {
            // Swap characters to create a new permutation
            swap(tiles, index, i);

            // Add the current permutation to the set
            uniqueSequences.add(new String(tiles, 0, index + 1));

            // Recursively generate permutations for the next index
            generatePermutations(tiles, index + 1, uniqueSequences);

            // Backtrack (swap back to restore the original array)
            swap(tiles, index, i);
        }
    }

    private void swap(char[] tiles, int i, int j) {
        char temp = tiles[i];
        tiles[i] = tiles[j];
        tiles[j] = temp;
    }
}